design/open/delta-diff-support.md

```diff
@@ -0,0 +1,179 @@
+1. UX: Be able to retrieve this from the lakeFS API, SDKs, CLI and UI
+
+### Defining Datasets
+
+Diffing Delta tables requires knowing where a Delta table is located within a repository.
+A single repository can potentially hold any number of Delta tables at different prefixes.
+
+To support this, we will add a new primitive to lakeFS: `Dataset`.
+
+A Dataset has 2 important properties:
+- id (`string`: a logical name for the dataset, e.g. `events`)
+- root location (`string`: path relative to the root of the repository)
+- Dataset type: (`enum{"Delta Lake", "Other"}`: representing the type of dataset at that location)
+- Description (`string`: arbitrary information about the dataset)
+
+By default, every repository has exactly 1 Dataset with the following attributes:
+
+- root location: `""`
+- Dataset type: `"Other"`
+
+This retains backwards compatibility with existing repositories.
+
+#### Datasets API
+
+`GET    /api/v1/repositories/<id>/datasets/` - list configured datasets for this repository
+`GET    /api/v1/repositories/<id>/datasets/<id>` - return the dataset location and type for the given path
+`DELETE /api/v1/repositories/<id>/datasets/<id>` - remove a dataset definition (does not remove data from the dataset)
+`PUT    /api/v1/repositories/<id>/datasets/<id>` - update an existing dataset setting
+`POST   /api/v1/repositories/<id>/datasets/` - configure a new dataset
+`GET    /api/v1/repositories/<id>/datasets/<id>/{leftRef}/diff/{rightRef}` - diff datasets
+
+##### Nesting
+
+To retain backwards compatibility, we are OK with datasets being nested within one another.
+For any given object or prefix, the Dataset they belong to will be the most specific one that includes the given path.
+
+Example:
+
+Say we have 2 datasets configured:
+
+- `{"id": "A", "root_location": "foo/", "type": "Other"}`
+- `{"id": "B", "root_location": "foo/bar/", "type": "Delta Lake"}`
+
+`path="foo/baz/x"`, will belong to `A` because it isn't covered by `B`
+
+`path="foo/bar/"`, will belong to `B` because it matches its `root_location`
+
+`path="foo/bar/baz/"`, will belong to `B` because it is the most specific covering dataset (longest `root_location` to match path)
+
+#### Diffing algorithms
+
+lakeFS will be able to use a different diff algorithm for every dataset type.
+
+The current diff API `/repositories/{repository}/refs/{leftRef}/diff/{rightRef}` will be extended and will be able to return a new `path_type`: `[added, removed, changed, conflict, prefix_changed, dataset]`
+
+```json
+{
+  "pagination": {
+    "has_more": true,
+    "next_offset": "string",
+    "results": 0,
+    "max_per_page": 0
+  },
+  "results": [
+    {
+      "type": "prefix_changed",
+      "path": "a/b/c/",
+      "path_type": "common_prefix",
+      "size_bytes": 0
+    }
+    {
+      "type": "prefix_changed",
+      "path": "foo/",
+      "path_type": "dataset", // <-- Notice the new path type
+      "size_bytes": 0
+    }
+  ]
+}
+```
```

Comment:
Apologies, I guess I should clarify this: 

What I meant is that format-dependent diffs *are* in fact a separate endpoint. If you simply call the existing diff API again on the `foo/` prefix in the example - you will get the current behavior which is an object-wise diff.

Reply:
Thanks, that makes a lot of sense!

Still a bit confusing.  I would expect the new endpoint to apply _only_ to the TLD of a single dataset.  Otherwise I can think of several very weird cases.